libkbfs/key_bundle_v3.go

```diff
@@ -109,28 +109,34 @@ func (udkimV3 UserDeviceKeyInfoMapV3) removeDevicesNotIn(
 	updatedUserKeys UserDevicePublicKeys) ServerHalfRemovalInfo {
 	removalInfo := make(ServerHalfRemovalInfo)
 	for uid, dkim := range udkimV3 {
+		userRemoved := false
 		deviceServerHalfIDs := make(deviceServerHalfRemovalInfo)
+		if _, hasUpdatedUser := updatedUserKeys[uid]; hasUpdatedUser {
+			for key, info := range dkim {
+				if !updatedUserKeys[uid][key] {
+					delete(dkim, key)
+					deviceServerHalfIDs[key] = append(
+						deviceServerHalfIDs[key],
+						info.ServerHalfID)
+				}
+			}
 
-		for key, info := range dkim {
-			if !updatedUserKeys[uid][key] {
+			if len(deviceServerHalfIDs) == 0 {
+				continue
+			}
+		} else {
+			userRemoved = true
+			for key, info := range dkim {
 				delete(dkim, key)
```

Comment:
Probably don't need to do this anymore since you're just deleting the whole `dkim` later.